sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java

```diff
@@ -382,4 +346,218 @@ ItemDeserializer getItemDeserializer() {
+                            return this.encryptionProcessor.initializeEncryptionSettingsAsync(true).then
+                                (Mono.defer(() -> replaceItemHelper(streamPayload, itemId, partitionKey, requestOptions,
+                                    itemClass, true)));
+                        }
+                    }
+                    return Mono.error(unwrappedException);
+                }));
+    }
+
+    public void setRequestHeaders(CosmosItemRequestOptions requestOptions) {
+        this.cosmosItemRequestOptionsAccessor.setHeader(requestOptions, Constants.IS_CLIENT_ENCRYPTED_HEADER, "true");
+        this.cosmosItemRequestOptionsAccessor.setHeader(requestOptions, Constants.INTENDED_COLLECTION_RID_HEADER, this.encryptionProcessor.getContainerRid());
+    }
+
+    public void setRequestHeaders(CosmosQueryRequestOptions requestOptions) {
+        this.cosmosQueryRequestOptionsAccessor.setHeader(requestOptions, Constants.IS_CLIENT_ENCRYPTED_HEADER, "true");
+        this.cosmosQueryRequestOptionsAccessor.setHeader(requestOptions, Constants.INTENDED_COLLECTION_RID_HEADER, this.encryptionProcessor.getContainerRid());
+    }
+
+    private <T> CosmosPagedFlux<T> queryItemsHelper(SqlQuerySpec sqlQuerySpec,
+                                                    CosmosQueryRequestOptions options,
+                                                    Class<T> classType,
+                                                    boolean isRetry) {
+        setRequestHeaders(options);
+        CosmosQueryRequestOptions finalOptions = options;
+        Flux<FeedResponse<T>>  tFlux = CosmosBridgeInternal.queryItemsInternal(container, sqlQuerySpec, options,
+            new Transformer<T>() {
+                @Override
+                public Function<CosmosPagedFluxOptions, Flux<FeedResponse<T>>> transform(Function<CosmosPagedFluxOptions, Flux<FeedResponse<JsonNode>>> func) {
+                    return queryDecryptionTransformer(classType, func);
+                }
+            }).byPage().onErrorResume(exception -> {
+            final Throwable unwrappedException = Exceptions.unwrap(exception);
+            if (unwrappedException instanceof CosmosException) {
+                final CosmosException cosmosException = (CosmosException) unwrappedException;
+                if (!isRetry && cosmosException.getStatusCode() == HttpConstants.StatusCodes.BADREQUEST &&
+                    cosmosException.getResponseHeaders().get(HttpConstants.HttpHeaders.SUB_STATUS)
+                        .equals(Constants.INCORRECT_CONTAINER_RID_SUB_STATUS)) {
+                    this.encryptionProcessor.getIsEncryptionSettingsInitDone().set(false);
+                    return this.encryptionProcessor.initializeEncryptionSettingsAsync(true).thenMany(
+                        (CosmosPagedFlux.defer(() -> queryItemsHelper(sqlQuerySpec,finalOptions, classType, true).byPage())));
```

Comment:
I am retrying the query one more time, another option is to fail the query and refresh cache for next attempt. Pros and cons below
1. Retry query internally will give user error free experience on collection recreate.
2. However if collection is recreated during the query drain , then we will have data from older collection , and again fresh query pipeline for the new container.
Thoughts ?

Reply:
1) if the collection is re-created prior to running the query, we should hide any failure from user and do retry internally.

2) if the collection creation happen during draining a query, then the behaviour is somewhat undefined. we probably should fail the query.

Reply:
We cannot differentiate between there two scenario   (collection recreate prior to query and collection recreate during the query) . Our current core sdk do the retry on 410/1000 (NAME_CACHE_IS_STALE) so we should retry here as well

Reply:
thanks for the explanation. if we are consistent with the core sdk behaviour, then we are good.